### ISSUE TITLE
Trigger

### ISSUE THREAD

**Issue by Systemsaholic** [lucas-filippi/portlight-groundhogg#1]
Is this where a Contact gets created in Groundhogg from a Ticket submit?

How does this get triggered?

Thanks

**Comment by lucas-filippi** [comment]
No, this plugin creates a contact in Groundhogg when SMS arrives and it's triggered [here](https://github.com/PortlightTechnologies/portlight-groundhogg/blob/1a2037540c75656097eb8a3c99ab6168c1aea19e/portlight-groundhogg.php#L14).

**Comment by Systemsaholic** [comment]
Right! duh!

Is there a reason you are hard coding the contact creation date?

'date_created'=> '2021-08-15 22:42:46',
'date_optin_status_changed'=> '2021-08-15 22:42:46'

Should that not be dynamic?


Al

**Comment by Systemsaholic** [comment]
> No, this plugin creates a contact in Groundhogg when SMS arrives and it's triggered [here](https://github.com/PortlightTechnologies/portlight-groundhogg/blob/1a2037540c75656097eb8a3c99ab6168c1aea19e/portlight-groundhogg.php#L14).

I still cannot get it to fire off?!

**Comment by lucas-filippi** [comment]
No reason for hard coded creation date, just did not have opportunity to improve it :).

You need to call a POST request to the root URL with From, Body and To not empty.

Paste your POST request here if you still have issues.

**Comment by Systemsaholic** [comment]
I'm sending requestes at /wp-admin/admin-ajax.php
action:wsdesk_api_create_ticket
api_key:72c492c45e7d03e97f3561a3be659d8c
phone_number:16137004540
request_description:Brand New Reply
request_email:16137003297@sms.com
request_title:Old Request

The POST is coming from Twilio when there is an Icoming SMS message.

**Comment by lucas-filippi** [comment]
Ok, the /wp-admin/admin-ajax.php request with action:wsdesk_api_create_ticket is to create a ticket in wsdesk, not to create a contact in Groundhogg.

This plugin (portlight-groundhogg) creates a contact in Groundhogg when you call a request to:
ULR: www.example.com
From: +18508766158
Body: SMS Test
To: +18508766111

And this is the request that is coming from Twilio when there is an Icoming SMS message, as you said.

**Comment by Systemsaholic** [comment]
The problem with that is they have seperate marketing SMS Number than the Support Number. Is there a way to triger the plugin when the WSdesk API is triggered instead? Like that we can contain the Support on the support side and the markeitng/Sales onthe the Groundhogg side. 

WSdesk API receives a call. Triggers the plugin that Queries the GH database for the matching Phone number. If there is no Match it creates the contact in GH. Else it pulls the Email Field from GH and fills the email Properties field in WSDesk.

**Comment by lucas-filippi** [comment]
Yes, it can be done in many ways. The simplest one that comes to my mind right now is to change [this trigger](https://github.com/PortlightTechnologies/portlight-groundhogg/blob/1a2037540c75656097eb8a3c99ab6168c1aea19e/portlight-groundhogg.php#L14) to expect the URL /wp-admin/admin-ajax.php with action wsdesk_api_create_ticket, etc.

**Comment by Systemsaholic** [comment]
But then it will only fire on an SMS ticket. We would need a way to get it to fire for both SMS, email and form submissions. So maybe it is better to have GH manage the request then pass it to WSDesk when the ['to'] field equals the Tech SUpport number.

**Comment by lucas-filippi** [comment]
Could you please describe in details the flow of SMS and form submission? From input to output. Including what is the plugin you use for form submission and where WSDesk is used.

Email is a bit tricky.

**Comment by Systemsaholic** [comment]
Lets start with this...

I think we need to check if a contact exists in the Groundhogg Database with query of the Mobile Numbers [HERE](https://github.com/PortlightTechnologies/portlight-groundhogg/blob/1a2037540c75656097eb8a3c99ab6168c1aea19e/portlight-groundhogg.php#L28)

If the contact exists we pull the email and name to insert in the WSdesk ticket. But we leave the from as the {mobile_phone}@sms.com to route wsdesk replies accordingly.

**Comment by Systemsaholic** [comment]
> Yes, it can be done in many ways. The simplest one that comes to my mind right now is to change [this trigger](https://github.com/PortlightTechnologies/portlight-groundhogg/blob/1a2037540c75656097eb8a3c99ab6168c1aea19e/portlight-groundhogg.php#L14) to expect the URL /wp-admin/admin-ajax.php with action wsdesk_api_create_ticket, etc.

What would that if statement look like? Just curious to try something...

**Comment by lucas-filippi** [comment]
Ok, if I understood correctly the flow is the following, correct me if I'm wrong:

1. SMS received
2. Create/Update Groundhogg contact 
3. Create/Update ticket in WSDesk

If so, steps 1 and 2 are already done. [This line](https://github.com/PortlightTechnologies/portlight-groundhogg/blob/1a2037540c75656097eb8a3c99ab6168c1aea19e/portlight-groundhogg.php#L25) query mobile number to check if the contact exists. If the contact exists, it adds a note with the SMS body, otherwise it creates a new contact.

For step 3, you can use [this plugin](https://github.com/PortlightTechnologies/portlight-wsdesk) as a template. It already creates a WSDesk ticket from incoming SMS.

**Comment by Systemsaholic** [comment]
> Ok, if I understood correctly the flow is the following, correct me if I'm wrong:
> 
> 1. SMS received
> 2. Create/Update Groundhogg contact
> 3. Create/Update ticket in WSDesk
> 
> If so, steps 1 and 2 are already done. [This line](https://github.com/PortlightTechnologies/portlight-groundhogg/blob/1a2037540c75656097eb8a3c99ab6168c1aea19e/portlight-groundhogg.php#L25) query mobile number to check if the contact exists. If the contact exists, it adds a note with the SMS body, otherwise it creates a new contact.
> 
> For step 3, you can use [this plugin](https://github.com/PortlightTechnologies/portlight-wsdesk) as a template. It already creates a WSDesk ticket from incoming SMS.

1. the SMS received has to be a Ticket and not a reply to a markeing SMS
2. The ticket created must be tied to a GH Contact based on their Mobile Phone number and not @carrier.com
3. The note should only be created when the ticket is closed and we save the conversation history.

**Comment by Systemsaholic** [comment]
New flow proposition

1. Webhook (SMS) received from Twilio
2. Check GH database for match based on Mobile Number
3. Create ticket in WSdesk database with GH and Twillio fields
4. When ticket is closed save conversation to GH contact's Notes.

**Comment by lucas-filippi** [comment]
1. It's done
2. It's already checking by mobile number, it just concatenates with @carrier.com because email is the key to search for a contact. Maybe there is a way to check only mobile number, need some analysis.
3. This can be done using the code from [this plugin](https://github.com/PortlightTechnologies/portlight-wsdesk).
4. I think for this one, you will need to change the WSDesk source code. Identify where the close ticket logic is and save the notes in GH contact.

**Comment by Systemsaholic** [comment]
2. 
SELECT * FROM wp_gh_contactmeta WHERE meta_value = $from
OR
SELECT contact_id FROM wp_gh_contactmeta WHERE meta_value = $from

**Comment by Systemsaholic** [comment]
1. but the way it works now we cannot destiguish between a Tech Issues and a Marketing Text Message.

We need a check on the ['to'] field and route the message accordingly.

**Comment by Systemsaholic** [comment]
3. Does not check for exiting ticket.... If its a reply. Need to query [HERE](https://github.com/PortlightTechnologies/portlight-wsdesk/blob/d8db1987a172be8ebb2c4227c243a605aaa9a122/portlight-wsdesk.php#L36) database and if exist set parent ticket ID.

**Comment by Systemsaholic** [comment]
4. We can do this easily by sending a webhook triggered by ticket status and GH can map the webhook to the Notes Field.

Not sure if we can amend the last not if the ticket get re-openned but that not a big deal

**Comment by Systemsaholic** [comment]
What do you think of this Lucas?

`//New ticket is created via HTTP POST Request to /wp-admin/admin.php From Twilio - Need to check if its in fact a SMS message
if (!empty($_POST['action']) && strpos($_POST['request_email'], '@sms') !== false) {
	
	//Include the WSDesk Functions
	include_once 'class-crm-ajax-functions.php';

	global $wpdb;


	//Check if the incoming SMS Number matches the Primary or the Mobile number of a contact in Groundhogg.
		
		//set temoprary email with SMS number@sms.com
		$requestEmail = $_POST['request_email'];
                //Strip the @sms.com from the incoming request_email field
		$contactMobile = preg_replace("/[^0-9]/", "",$_POST['request_email']);
	   	//Get the GH contact ID where the Mobile Number matches in the database
		$contactId = $wpdb->get_results("SELECT contact_id FROM wp_gh_contactmeta WHERE meta_value = '$contactMobile' LIMIT 1" );
		//Get the new ticket ID
		$ticketId = $wpdb->get_results("SELECT ticket_id FROM wp_wsdesk_tickets WHERE ticket_email = '$requestEmail' LIMIT 1");
		

	//IF the contact exists - then pull the contact's Full Name and email to insert into the WSDESK ticket

	if (!empty($contactId)){
		
		//gather the contact details from GH
		$contactFName = $wpdb->get_results("SELECT first_name FROM wp_gh_contacts WHERE id = '$contactId'" );
		$contactLName = $wpdb->get_results("SELECT last_name FROM wp_gh_contacts WHERE id = '$contactId'" );
		$contactEmail = $wpdb->get_results("SELECT email FROM wp_gh_contacts WHERE ID = '$contactId'" );
		
		//update the CRM Ticket Meta Fields
		eh_crm_update_ticketmeta($ticketId, 'field_MK09', $contactEmail);
		eh_crm_update_ticketmeta($ticketId, 'field_PF91', $contactMobile);

	} else {
		//Else (no match on phone number) - Then create the contact in groundhogg with made up email PhoneNumber@sms.com
		$wpdb->insert('wp_gh_contacts', array(
			'email' => $_POST['request_email'],
			'user_id' => 1,
			'owner_id' => 1,
			'optin_status' => 1,
			'date_created' => date('Y-m-d H:i:s'),
			'date_optin_status_changed' => date('Y-m-d H:i:s')
		));
	};
}`

**Comment by lucas-filippi** [comment]
Sounds good, left some comments in your PR.
